Proper-command-retry-handling.md

```diff
@@ -0,0 +1,35 @@
+* Solve [ISPN-3918](https://issues.jboss.org/browse/ISPN-3918) - although this was amongst the goals initially, the fix would have negative performance impact or would be out of scope (flag-affected behaviour).
+* Fix anything related to functional events
+* Deal with delta-awares more than what is required to pass current testsuite
+
+# Concepts
+* The fact that a command has been applied to a node is stored as `InvocationRecord` in metadata.
+  * Invocation records expire over time; this is executed by the expiration thread. The default expiration time is 4 times the replication timeout, but if an attempt is made to apply the value again, the expiration is postponed.
+  * When a command is completed on the originator, the (key, command-invocation-id) pair is sent to background process (in `InvocationManager`) which invalidates he invocation records in a batch.
+  * As the invocation record has to be held in metadata even for removed entries, both data container and cachestores have to be able to hold entries with null value (aka tombstones).
+  * Since backup may not know previous value of an entry when the command is applied (replace is non-conditional on backup owner when it succeeds on primary and does not have to load previous value), the return value on primary has to be stored in the invocation record for the case that this backup becomes the next primary. From reasons above, primary has to replicate return value of a command to all backups, too.
+* `InvocationManager` keeps information about invalidated invocations per segment, and once this exceeds a threshold (for given segment) - 100 by default, customizable through `infinispan.invocations` property ATM - sends `ForgetInvocationsCommand` to all write owners of given segment. Non-current owners are expected to drop the invocation records when these lose ownership of the segment.
+  * In async modes we can't do any invalidations because originator does not know when the command finishes; we have to rely on expiration. However, we have to store the invocation record even if the entry is committed with `ctx.isOriginLocal()`. Regrettably this forces us to store the information about command being executed in sync/async mode in the command itself.
+* Commands that require the same previous value on backup as on primary implement `StrictOrderingCommand`. At this point these are ReadWrite* functional commands. When a strict command is replicated to backup, primary adds command invocation id of previous command (or null if there's no previous record). The backup checks its last record against this last invocation id and applies the command only in case of match. In other case it retrieves full entry including metadata with invocation records from the primary owner, and stores this value (potentially applying the command it it was not applied - this depends on classical/triangle routing algorithms).
```

Comment:
Seems like this would get rid of `ReadWriteKeyValueCommand` shipping `prevValue` and `prevMetadata` around? Your approach would appear to make the command more lightweight, and instead use the command id of previous command affecting that key/value pair.

However, if no `prev*` is shipped, you'd have to get `prev*` data from the `InvocationManager` on demand, right?

It'd be good to explain this a bit further, since from my POV, one of the key reasons why functional API is not yet graduated to not being experimental is due to the cost required to deal with retries. As I've explained before, if you have a List of 1000 elements and you want to pass a lambda around to add an element in the middle of the list, having to ship around the previous value defeats the purpose of functional cmds.